### PR TITLE
update msal-browser caching doc

### PR DESCRIPTION
**Pull request by derisen**
- Expands the current doc to answer some FAQs
- Mentions `memoryStorage` option as available

### CODE REVIEW

File: lib/msal-browser/docs/caching.md [hunk 1]
```diff
@@ -1,3 +1,88 @@
...
+
+`sessionStorage` is more secure, but `localStorage` allows application or application instances on the same domain to share cache. This enables single sign-on between tabs and user sessions.
+
+Additionally, you can opt-out of browser storage by setting the `cacheLocation` to `memoryStorage` in msal configuration. This means that your application will lose authentication state upon page refresh (however, since users will still have an active session with the IdP, they would be able to re-authenticate without any prompts).
+
+We encourage you to explore the options and make the best decision for your application.
+
+### Cookie storage
+
+MSAL Browser can be configured to use cookies for storing the authentication state, in addition to browser or memory storage. This is useful for apps that need to support older browsers such as Internet Explorer. Please refer to [configuration](./configuration.md#cache-config-options) for more.
```
Comment (jasonnutter) [inline lib/msal-browser/docs/caching.md]: Should be made clear that tokens are not stored in cookies.

File: lib/msal-browser/docs/caching.md [hunk 2]
```diff
@@ -1,3 +1,88 @@
...
+```
+
+## Cache storage
+
+By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of storage:
+
+- `sessionStorage` (default):
+- `localStorage`
+
+`sessionStorage` is more secure, but `localStorage` allows application or application instances on the same domain to share cache. This enables single sign-on between tabs and user sessions.
```
Comment (jasonnutter) [inline lib/msal-browser/docs/caching.md]: "more secure" is a bit of a misnomer. Instead, it should be noted that session storage is not shared between tabs and is cleared when the tab is closed. Probably best to have a table to compares localstorage/sessionstorage/memorystorage as to when the cache is cleared (persistent, on tab close, on page refresh/navigation, etc), whether it is shared between tabs, and whether redirect operations are supported (not supported with memory storage).

Comment (jasonnutter) [reply lib/msal-browser/docs/caching.md]: Also worth calling out that temporary cache entries will always be stored in session storage or in memory, and that msal will fallback to memory storage if local/session storage are not available.

File: lib/msal-browser/docs/caching.md [hunk 3]
```diff
@@ -1,3 +1,88 @@
...
+
+// Create the msal application object
+const pca = new msal.PublicClientApplication({
+    auth: {
+        clientId: "Enter_the_Application_Id_Here", // e.g. "b1b60dca-c49d-496e-9851-xxxxxxxxxxxx" (guid)
+        authority: "https://login.microsoftonline.com/Enter_the_Tenant_Info_Here", // e.g. "common" or your tenantId (guid),
+        redirectUri: "/"
+    },
+    cache: {
+       cacheLocation: "sessionStorage"
```
Comment (jasonnutter) [inline lib/msal-browser/docs/caching.md]: Can we show use the enum?

File: lib/msal-browser/docs/caching.md [hunk 4]
```diff
@@ -1,3 +1,88 @@
...
+By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of storage:
+
+- `sessionStorage` (default):
+- `localStorage`
+
+`sessionStorage` is more secure, but `localStorage` allows application or application instances on the same domain to share cache. This enables single sign-on between tabs and user sessions.
+
+Additionally, you can opt-out of browser storage by setting the `cacheLocation` to `memoryStorage` in msal configuration. This means that your application will lose authentication state upon page refresh (however, since users will still have an active session with the IdP, they would be able to re-authenticate without any prompts).
+
+We encourage you to explore the options and make the best decision for your application.
```
Comment (jasonnutter) [inline lib/msal-browser/docs/caching.md]: Can probably remove this sentence.

File: lib/msal-browser/docs/caching.md [hunk 5]
```diff
@@ -1,3 +1,88 @@
...
+        redirectUri: "/"
+    },
+    cache: {
+       cacheLocation: "sessionStorage"
+    }
+});
+
+/**
+* login* and acquireToken* APIs return an account object containing "homeAccountId"
+* you should keep a record of this in your app and use it later on when calling acquireTokenSilent
```
Comment (jasonnutter) [inline lib/msal-browser/docs/caching.md]: Instead, should mention using `setActiveAccount`.

Comment (jasonnutter) [reply lib/msal-browser/docs/caching.md]: Also, the code shippets below probably aren't needed for this doc.

File: lib/msal-browser/docs/caching.md [hunk 6]
```diff
@@ -1,3 +1,88 @@
...
+    - interaction status
+- Telemetry
+    - previous failed request 
+    - performance data
+
+## Remarks
+
+- MSAL Browser's cache schema is compatible with other MSALs. However, there is no cache sharing support at the moment. 
+- We do not recommend apps having business logic dependent on direct use of entities in the cache. Instead, use the appropriate MSAL API when you need to acquire tokens or retrieve accounts.
+- Keys used to encrypt proof of possession (PoP) tokens are stored using the [IndexedDB API](https://developer.mozilla.org/docs/Web/API/IndexedDB_API). For more information, please refer to [access-token-proof-of-possession](./access-token-proof-of-possession.md#pop-key-management).
```
Comment (jasonnutter) [inline lib/msal-browser/docs/caching.md]: POP will use memory storage and IDB together.

File: lib/msal-browser/docs/caching.md [hunk 7]
```diff
@@ -1,3 +1,88 @@
-## Cache Storage
+# Caching
 
-We offer two methods of storage for Msal, `localStorage` and `sessionStorage`.  Our recommendation is to use `sessionStorage` because it is more secure in storing tokens that are acquired by your users, but `localStorage` will give you Single Sign On accross tabs and user sessions.  We encourage you to explore the options and make the best decision for your application.
+When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. APIs like `acquireTokenSilent()` retrieves tokens from the cache for a given account.
```
Comment (sameerag) [inline lib/msal-browser/docs/caching.md]: Can we list all the APIs that retrieve cached tokens?

Comment (derisen) [reply lib/msal-browser/docs/caching.md]: Just added. Would that work?

File: lib/msal-browser/docs/caching.md (hunk 6 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: revised

File: lib/msal-browser/docs/caching.md (hunk 5 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: removed the snippet

File: lib/msal-browser/docs/caching.md (hunk 4 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: done

File: lib/msal-browser/docs/caching.md (hunk 3 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: done

File: lib/msal-browser/docs/caching.md (hunk 1 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: done

File: lib/msal-browser/docs/caching.md (hunk 2 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: revised and added a table

File: lib/msal-browser/docs/caching.md [hunk 8]
```diff
@@ -1,3 +1,78 @@
-## Cache Storage
+# Caching in MSAL
 
-We offer two methods of storage for Msal, `localStorage` and `sessionStorage`.  Our recommendation is to use `sessionStorage` because it is more secure in storing tokens that are acquired by your users, but `localStorage` will give you Single Sign On accross tabs and user sessions.  We encourage you to explore the options and make the best decision for your application.
+When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. The following MSAL APIs will perform one or more cache lookups to acquire tokens or retrieve accounts:
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]: Technically, all of our APIs perform cache lookups but only `acquireTokenSilent` returns tokens from the cache. IMO this section should only mention `acquireTokenSilent`

File: lib/msal-browser/docs/caching.md [hunk 9]
```diff
@@ -1,3 +1,78 @@
...
+        authority: "https://login.microsoftonline.com/Enter_the_Tenant_Info_Here", // e.g. "common" or your tenantId (guid),
+        redirectUri: "/"
+    },
+    cache: {
+       cacheLocation: BrowserCacheLocation.SessionStorage // "sessionStorage"
+    }
+});
+```
+
+By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of storage: `sessionStorage` (default) and `localStorage`. In addition, MSAL provides `memoryStorage` option which allows you to opt-out of storing the cache in browser storage.
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]:
```suggestion
By default, MSAL stores the various authentication artifacts it obtains from the IdP in browser storage using the [Web Storage API](https://developer.mozilla.org/docs/Web/API/Web_Storage_API) supported by all modern browsers. Accordingly, MSAL offers two methods of persistent storage: `sessionStorage` (default) and `localStorage`. In addition, MSAL provides `memoryStorage` option which allows you to opt-out of storing the cache in browser storage.
```

File: lib/msal-browser/docs/caching.md [hunk 10]
```diff
@@ -1,3 +1,78 @@
...
+|------------------|-------------------------|-----------------------------|-------------------------|
+| `sessionStorage` | window/tab close        | No                          | Yes                     |
+| `localStorage`   | user logout             | Yes                         | Yes                     |
+| `memoryStorage`  | page refresh/navigation | No                          | No                      |
+
+> :bulb: While the authentication state may be lost in session and memory storage due to window/tab close or page refresh/navigation, respectively, users will still have an active session with the IdP as long as the session cookie is not expired and might be able to re-authenticate without any prompts.
+
+### Cookie storage
+
+MSAL Browser can be configured to use cookies for storing the authentication state. This is useful for apps that need to support older browsers such as Internet Explorer. Note that when this option is choosed, tokens themselves are still stored in browser or memory storage. Please refer to [configuration](./configuration.md#cache-config-options) for more.
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]:
```suggestion
MSAL Browser can be configured to use cookies for storing temporary authentication artifacts. This is useful for apps that need to support older browsers such as Internet Explorer. Note that when this option is chosen, tokens themselves are still stored in browser or memory storage. Please refer to [configuration](./configuration.md#cache-config-options) for more.
```

File: lib/msal-browser/docs/caching.md [hunk 11]
```diff
@@ -1,3 +1,78 @@
...
+    - previous failed request 
+    - performance data
+
+> :bulb: Temporary cache entries will always be stored in session storage or in memory. MSAL will fallback to memory storage if local/session storage is not available.
+
+> :bulb: The authorization code is only stored in memory and will be discarded after redeeming it for tokens.
+
+## Remarks
+
+- MSAL Browser's cache schema is compatible with other MSALs. However, there is no cache sharing support at the moment.
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]: This can be removed as it's not relevant at this time

Comment (derisen) [reply lib/msal-browser/docs/caching.md]: Sure thing, removed

File: lib/msal-browser/docs/caching.md (hunk 8 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: Ah, very well, reverted this

File: lib/msal-browser/docs/caching.md [hunk 12]
```diff
@@ -1,3 +1,70 @@
-## Cache Storage
+# Caching in MSAL
 
-We offer two methods of storage for Msal, `localStorage` and `sessionStorage`.  Our recommendation is to use `sessionStorage` because it is more secure in storing tokens that are acquired by your users, but `localStorage` will give you Single Sign On accross tabs and user sessions.  We encourage you to explore the options and make the best decision for your application.
+When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. APIs like `acquireTokenSilent()` retrieves access tokens from the cache for a given account.
```
Comment (tnorling) [inline lib/msal-browser/docs/caching.md]:
```suggestion
When MSAL acquires a token, it caches it for future usage. MSAL manages token lifetimes and refreshing for you. The `acquireTokenSilent()` API retrieves access tokens from the cache for a given account and renews them if needed.
```

**Review by tnorling — APPROVED** [review]

File: lib/msal-browser/docs/caching.md [hunk 13]
```diff
@@ -1,3 +1,70 @@
...
+|------------------|-------------------------|-----------------------------|-------------------------|
+| `sessionStorage` | window/tab close        | No                          | Yes                     |
+| `localStorage`   | user logout             | Yes                         | Yes                     |
+| `memoryStorage`  | page refresh/navigation | No                          | No                      |
+
+> :bulb: While the authentication state may be lost in session and memory storage due to window/tab close or page refresh/navigation, respectively, users will still have an active session with the IdP as long as the session cookie is not expired and might be able to re-authenticate without any prompts.
+
+### Cookie storage
+
+MSAL Browser can be configured to use cookies for storing temporary authentication artifacts. This is useful for apps that need to support older browsers such as Internet Explorer. Note that when this option is chosen, tokens themselves are still stored in browser or memory storage. Please refer to [configuration](./configuration.md#cache-config-options) for more.
```
Comment (peterzenz) [inline lib/msal-browser/docs/caching.md]: Isn't internet explorer officially end of life at this point?  What still supported browsers **require** this?  If we can suggest folks not use cookies that would be my preference.

File: lib/msal-browser/docs/caching.md [hunk 14]
```diff
@@ -1,3 +1,70 @@
...
+
+### Security
+
+We consider session/local storage secure as long as your application does not have cross-site scripting (XSS) and related vulnurabilities. Please refer to the [OWASP XSS Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) for securing your applications against XSS. If you are still concerned, we recommend using the `memoryStorage` option instead.
```
Comment (peterzenz) [inline lib/msal-browser/docs/caching.md:40]: Do we need to flag the OWASP page as being an external link?

File: lib/msal-browser/docs/caching.md [hunk 15]
```diff
@@ -1,3 +1,70 @@
...
+
+## Cached artifacts
+
+To faciliate efficient token acquisition while maintaining a good UX, MSAL caches various artifacts resulting from its API calls. At a glance, these are:
+
+- Persistent artifacts
```
Comment (peterzenz) [inline lib/msal-browser/docs/caching.md]: What does persistent mean here?  AT/IDToken might last for an hour, is that temporary?  How does that compare to the definition of Temporary on line 51 below.

**Review by peterzenz — APPROVED** [review]

File: lib/msal-browser/docs/caching.md (hunk 13 above)
Comment (jasonnutter) [reply lib/msal-browser/docs/caching.md]: I would rephrase this to specifically note that it is needed for browsers that may clear local/session storage during redirect-based login, which includes IE and Firefox Inprivate.

Comment (jasonnutter) [reply lib/msal-browser/docs/caching.md]: (as opposed to saying this is needed only for older browsers)

File: lib/msal-browser/docs/caching.md (hunk 15 above)
Comment (jasonnutter) [reply lib/msal-browser/docs/caching.md]: Persistent would be anything that lasts for longer than the duration of the request (i.e. tokens, accounts)

File: lib/msal-browser/docs/caching.md (hunk 13 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: revised!

File: lib/msal-browser/docs/caching.md (hunk 15 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md]: reworded this section and added a link to token lifetimes doc in case if there's confusion

File: lib/msal-browser/docs/caching.md (hunk 14 above)
Comment (derisen) [reply lib/msal-browser/docs/caching.md:40]: Hmm, I would say orgs like OWASP and MDN need no special mentioning, but if we were to flag these, what would that look like? :/
